### PR TITLE
Fix evaluation order for overapplications when using Closure

### PR DESCRIPTION
**Pull request by mshinwell**
When using the Closure pass (i.e. non-Flambda) the evaluation order for overapplications does not match the bytecode compiler and Flambda: as shown by PR#6136, the subset of arguments forming the full application are evaluated first, then the function is called; and then the remaining arguments are evaluated.  This patch fixes that by let-binding the arguments before the call.

### CODE REVIEW

**Comment by alainfrisch** [comment]
> This patch fixes that by let-binding the arguments before the call.

It would be useful to see the impact on performance, even on some micro-benchmarks only.  Evaluating extra arguments before the call is likely to create more transit with the memory.

File: Changes [hunk 1]
```diff
@@ -185,6 +185,9 @@ Next version (4.05.0):
   by zero.
   (Jeremy Yallop)
 
+- GPR#967, PR#6536: Fix Closure so that overapplication evaluation order
+  matches the bytecode compiler and Flambda.
```
Comment (yallop) [inline Changes:236]: The Mantis PR number should be 6136, not 6536.  It's also conventional to list the author and the reporter.

**Comment by xavierleroy** [comment]
Over-applications are exceedingly rare in practice, so I'm not too worried about a performance degradation here.

**Comment by mshinwell** [comment]
I also think the rarity of these cases means that we shouldn't really be worrying about the performance impact of this change.  Furthermore, constructing benchmarks that actually give meaningful results for changes like this probably isn't entirely straightforward, and given limited time I think our attention is best concentrated elsewhere (e.g. #966).

Would anyone like to comment on the correctness of this patch?

**Comment by mshinwell** [comment]
@let-def I nominate you to review. :)

**Comment by let-def** [comment]
This looks good to me.

I was wondering if there was an interest in forcing evaluation order of all arguments instead of only over-applied ones (folding only on rem_args), but I don't think this change anything in practice.

**Comment by mshinwell** [comment]
Modulo any problems in the backend, of course, but we're addressing those in a different GPR.  Thanks for the review, I will merge this.

**Pull request merged**